File: rust/core-types/src/array_descriptor.rs
use crate::{NumericalType, StridedLayout};

#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub struct ArrayDescriptor {
	layout: StridedLayout,
	data_type: NumericalType,
}

impl Default for ArrayDescriptor {
	fn default() -> Self {
		Self {
			layout: StridedLayout::default(),
			data_type: NumericalType::Unknown,
		}
	}
}

impl ArrayDescriptor {
	pub fn new(layout: StridedLayout, data_type: NumericalType) -> Self {
		Self { layout, data_type }
	}

	pub fn layout(&self) -> &StridedLayout {
		&self.layout
	}

	pub fn data_type(&self) -> NumericalType {
		self.data_type
	}
}

pub fn is_initialized(descriptor: &ArrayDescriptor) -> bool {
	descriptor.data_type() != NumericalType::Unknown
}

pub fn compute_storage_requirement(descriptor: &ArrayDescriptor) -> usize {
	descriptor.layout().compute_storage_requirement()
		* descriptor.data_type().size_bytes().unwrap_or(0)
}

File: src/core/ndarray/array_descriptor.cpp
// SPDX-License-Identifier: GPL-3.0-only

#include <xmipp4/core/ndarray/array_descriptor.hpp>

#include <boost/functional/hash.hpp>

namespace xmipp4 
{
namespace ndarray
{

array_descriptor::array_descriptor() noexcept
	: m_data_type(numerical_type::unknown)
{
}

array_descriptor::array_descriptor(
	layout::strided_layout layout,
	numerical_type data_type
) noexcept
	: m_layout(std::move(layout))
	, m_data_type(data_type)
{
}

array_descriptor::array_descriptor(const array_descriptor &other) = default;
array_descriptor::array_descriptor(array_descriptor &&other) noexcept = default;
array_descriptor::~array_descriptor() = default;

array_descriptor&
array_descriptor::operator=(const array_descriptor &other) = default;
array_descriptor&
array_descriptor::operator=(array_descriptor &&other) noexcept = default;

std::size_t array_descriptor::hash() const noexcept
{
	auto seed = m_layout.hash();
	boost::hash_combine(seed, boost::hash_value(m_data_type));
	return seed;
}

const layout::strided_layout& array_descriptor::get_layout() const noexcept
{
	return m_layout;
}

numerical_type array_descriptor::get_data_type() const noexcept
{
	return m_data_type;
}

bool operator==(
	const array_descriptor &lhs, 
	const array_descriptor &rhs
) noexcept
{
	return 
		lhs.get_data_type() == rhs.get_data_type() &&
		lhs.get_layout() == rhs.get_layout();
}

bool operator!=(
	const array_descriptor &lhs, 
	const array_descriptor &rhs
) noexcept
{
	return !(lhs == rhs);
}

array_descriptor make_contiguous_array_descriptor(
	span<const std::size_t> extents,
	numerical_type data_type
)
{
	return array_descriptor(
		layout::strided_layout::make_contiguous_layout(extents),
		data_type
	);
}

bool is_initialized(const array_descriptor &descriptor) noexcept
{
	return descriptor.get_data_type() != numerical_type::unknown;
}

std::size_t compute_storage_requirement(const array_descriptor &descriptor)
{
	const auto &layout = descriptor.get_layout();
	const auto data_type = descriptor.get_data_type();
	return layout.compute_storage_requirement() * get_size(data_type);
}

} // namespace ndarray
} // namespace xmipp4
